Add support for KHR_materials_unlit

OmarShehata opened the pull request:
Resolves https://github.com/AnalyticalGraphicsInc/cesium/issues/6507.

As described in the spec, any glTF that has this extension defined as shown below will be rendered just using its base color.

```
"materials": [
    {
        "name": "unlit_with_fallback",
        "pbrMetallicRoughness": {
            "baseColorFactor": [ 1.0, 1.0, 1.0, 1.0 ],
            "baseColorTexture": { "index": 0 },
            "roughnessFactor": 0.9,
            "metallicFactor": 0.0
        },
        "extensions": {
            "KHR_materials_unlit": {}
        }
    }
]
```

I implemented this by simply setting `hasNormals` and `hasTangents`, which generates a simple shader with no lighting. 

Here is Suzanne with PBR. 

![before](https://user-images.githubusercontent.com/1711126/44868866-12f6d300-ac5a-11e8-9a9b-fe2cf6473358.png)

And with the unlit extension. 

![after](https://user-images.githubusercontent.com/1711126/44868876-1db16800-ac5a-11e8-9894-49fcae9e1f33.png)

Unlit materials are affected by shadows. This is `viewer.shadows = true`. 

![after_shadows](https://user-images.githubusercontent.com/1711126/44868888-24d87600-ac5a-11e8-9a99-186e9caea546.png)

@lilleyse any feedback is welcome! This is the current shader that gets generated:

```
void main(void) 
{
    vec4 baseColorWithAlpha = u_baseColorFactor;
    vec3 baseColor = baseColorWithAlpha.rgb;
    vec3 color = baseColor;
    color += u_emissiveFactor;
    color = LINEARtoSRGB(color);
    gl_FragColor = vec4(color, 1.0);
}
```

* Should it be using emissive factor?  
* Are there other potential code paths to test? 
* I couldn't find any tests for `processPbrMetallicRoughness` in this branch.

cesium-concierge commented:
Thanks for the pull request @OmarShehata!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

emackey commented:
@OmarShehata Unlit materials should not apply emissive.  https://github.com/KhronosGroup/glTF/issues/1391

emackey commented:
Also `KHR_materials_unlit` should be added to the list of extensions allowed to appear in `extensionsRequired`.  See `ModelUtility.checkSupportedExtensions`.

OmarShehata commented:
Thanks for the link @emackey ! I changed it to ignore emissive and occlusion as suggested there. The new shader looks like:

```
void main(void) 
{
    vec4 baseColorWithAlpha = u_baseColorFactor;
    vec3 baseColor = baseColorWithAlpha.rgb;
    vec3 color = baseColor;
    color = LINEARtoSRGB(color);
    gl_FragColor = vec4(color, 1.0);
}
```

> Also KHR_materials_unlit should be added to the list of extensions allowed to appear in extensionsRequired. See ModelUtility.checkSupportedExtensions.

It should [already be in there](https://github.com/AnalyticalGraphicsInc/cesium/pull/6977/commits/ccad16c02266c3a8ce3cf252d80b8972a1b31756) !

ggetz commented:
@OmarShehata There's no explcit tests for the `processX` functions as it was just moved out of gltf-pipeline, but `ModelSpec.js` tests for rendering pbr materials.

emackey commented:
@OmarShehata This looks good to me.  One last thing to be aware of, currently Cesium is known to not be following the glTF spec which calls for calculating missing normals (issue #6506).  If/when that bug gets fixed, we should make sure that normals are not calculated for unlit materials.

OmarShehata commented on Source/Scene/processPbrMetallicRoughness.js:
```diff
@@ -660,6 +670,8 @@ define([
         }
         fragmentShader += '}\n';
 
+        console.log(fragmentShader);
```
Yes, sorry. Thanks for catching that!

OmarShehata commented:
Thanks for the review Matt! @ggetz I added a test there to verify that it correctly renders the base color when using the unlit extension. I added a new glTF to test on. Since it was just a modified `Box` I put it in the same directory but let me know if that should go somewhere else.

ggetz commented on Specs/Scene/ModelSpec.js:
```diff
@@ -2692,6 +2693,21 @@ defineSuite([
         });
     });
 
+    fit('renders with the unlit extension', function() {
```
Remove the `fit`

ggetz commented on Specs/Scene/ModelSpec.js:
```diff
@@ -120,6 +120,7 @@ defineSuite([
     var boomBoxUrl = './Data/Models/PBR/BoomBox/BoomBox.gltf';
     var boxPbrUrl = './Data/Models/PBR/Box/Box.gltf';
+    var boxPbrUnlitUrl = './Data/Models/PBR/Box/Box-Unlit.gltf';
     var boxAnimatedPbrUrl = './Data/Models/PBR/BoxAnimated/BoxAnimated.gltf';
```
Put the model in it's own folder like we do here. Also use camel case like we do for the other models. The path should be `./Data/Models/PBR/BoxUnlit/BoxUnlit.gltf`. Just for consistency's sake.

OmarShehata commented:
I also tested this with a base texture instead of a base color and it seems to correctly render the unlit texture. Here's the produced shader in that case.

```
void main(void) 
{
    vec4 baseColorWithAlpha = SRGBtoLINEAR4(texture2D(u_baseColorTexture, v_texcoord_0));
    baseColorWithAlpha *= u_baseColorFactor;
    vec3 baseColor = baseColorWithAlpha.rgb;
    vec3 color = baseColor;
    color = LINEARtoSRGB(color);
    gl_FragColor = vec4(color, 1.0);
}
```

emackey commented:
Base texture shader logic looks good, thanks.

OmarShehata commented:
@ggetz I've just verified that the CI errors in this branch are also erroring in the base branch. It's also peculiar because the test I added passes when ran individually but fails when ran together with the other tests in `ModelSpec` (I tried this with a few other tests in that file, so it's not just mine).

Do you know why/is this expected?

lilleyse commented:
@OmarShehata the model tests can be a little weird at times and you may have missed resetting a model's show or something like that. Look at what other tests do to reset the state after they are finished.

ggetz commented:
@OmarShehata There's an error in a previous test for `Model.js` in that branch which causes all subsequent tests to fail. I saw your test pass individually so you should be fine.

emackey commented on CHANGES.md:
```diff
@@ -10,6 +10,7 @@ Change Log
 * Improved support for polygon entities using `perPositionHeight`, including supporting vertical polygons.  This also improves KML compatibility. [#6791](https://github.com/AnalyticalGraphicsInc/cesium/pull/6791)
 * Added `Cartesian3.midpoint` to compute the midpoint between two `Cartesian3` positions [#6836](https://github.com/AnalyticalGraphicsInc/cesium/pull/6836)
 * Added `equalsEpsilon` methods to `OrthographicFrustum`, `PerspectiveFrustum`, `OrthographicOffCenterFrustum` and `PerspectiveOffCenterFrustum`.
+* Added support for glTF extension [KHR_materials_unlit](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_unlit) [#6977](https://github.com/AnalyticalGraphicsInc/cesium/pull/6977).
```
Looks like this isn't targeting master, as the changelog is 2 versions out of date as of today's release.

OmarShehata replied:
Yeah I'm merging into https://github.com/AnalyticalGraphicsInc/cesium/pull/6805 since that PR moves `gltfPipeline` from `ThirdParty` into Cesium core.

ggetz replied:
@OmarShehata Create a section for October's release and move your changes there. Thanks!

OmarShehata replied:
Oh, I see the problem now. I moved it into a new section for October. Thanks for pointing this out!

lilleyse commented:
Took a quick look as well, looks good!
